### PR TITLE
[MERGE][REF] web: new RelationalModel

### PR DESCRIPTION
**Pull request by aab-odoo**
Before this PR, we had several implementations of a relational
model. The form view used the BasicRelationalModel, which was a
wrapper around the BasicModel. The list view used the
RelationalModel (a simpler version that didn't support complex
usecases of x2many fields). The kanban view used the KanbanModel,
an extension of RelationalModel.

This PR removes all these, and replaces them by a single model.
It is called RelationalModel, and as a similar structure as its
predecessor (with datapoints). It now handles tricky x2many usecases
of the form view though.

This PR also comes with several important changes:
 1) The new model is based on unity reads [1]. This means that all
 data required by the view is fetched at once, and datapoints are
 created with their data (whereas before, they were responsible to
 fetch their data). There's an exception for special data: they are
 no longer handler by the model. A hook "useSpecialData" has been
 created, and it allows fields (components) to fetch additional
 data they require.

 2) The new model uses onchange2 [2]. With onchange2, we send and
 receive a diff, instead of the whole state of the record. Morever,
 onchange2 uses unity, which means that returned commands for x2manys
 (e.g. command 4, LINK) contain the values of the corresponding
 records, so we don't need to fetch them in a separate call
 afterwards.

 3) The new model no longer triggers deep renderings (render(true)).
 It is instead based on fine-grained reactivity (i.e. only what
 depends on what has changed is re-rendered). We introduced a hook
 "useRecordObserver" which allows a component to subscribe itself to
 some keys in a record datapoint, and register a callback to
 execute when those keys change (e.g. to fetch data).

[1] odoo/odoo@7d2baaa0c726a7b0dd4fe7226862f960c9c59b73
[2] odoo/odoo@f5e6494da3aedb545ead34d0d7f4b1b86499cb7f

Task~3179751

Co-authored-by: Aaron Bohy <aab@odoo.com>
Co-authored-by: FrancoisGe <fge@odoo.com>
Co-authored-by: Jorge Pinna Puissant <jpp@odoo.com>
Co-authored-by: Pierre Rousseau <pro@odoo.com>

### CODE REVIEW

**Comment by pro-odoo** [comment]
Hello @odoo/rd-security, can we have a review ? Here are the changes:

- [odoo/addons/web/static/src/model/relational_model/record.js:219](https://github.com/odoo/odoo/blob/d9fdc1495c26a20995ef66dcbe2224df6c9eeeef/addons/web/static/src/model/relational_model/record.js#L219)
```markup(<ul>${items.join("")}</ul>),```

This markup is used to display fields name in a notification, `items` is a list of `<li/>` elements, with the field name escaped.

- [odoo/addons/web/static/src/model/relational_model/utils.js:338](https://github.com/odoo/odoo/blob/d9fdc1495c26a20995ef66dcbe2224df6c9eeeef/addons/web/static/src/model/relational_model/utils.js#L338)
```return markup(value || "");```

The value here is the value of an html field received from the server

- [odoo/addons/web/static/tests/views/list_view_tests.js:5487](https://github.com/odoo/odoo/blob/d9fdc1495c26a20995ef66dcbe2224df6c9eeeef/addons/web/static/tests/views/list_view_tests.js#L5487)
```props.body = markup(```

Used in a test, to ensure that `markup` is allowed in a dialog. The code here is simply formatted. See the original PR: https://github.com/odoo/odoo/pull/123532

Many thanks !

**Review by xmo-odoo — COMMENTED** [review]
@robodoo override=ci/security

File: addons/survey/static/src/question_page/question_page_one2many_field.js [hunk 1]
```diff
@@ -27,6 +27,28 @@ patch(X2ManyFieldDialog.prototype, 'survey_question_chaining_with_validation', {
...
+ * For convenience, we'll prevent closing the question form dialog and
+ * stay in edit mode to make sure only valid records are saved. Therefore,
+ * in case of error occurring when saving we will replace default error
+ * modal with a notification.
+ */
+
+class SurveySaveError extends Error {}
+function SurveySaveErrorHandler(env, error, originalError) {
+    if (originalError instanceof SurveySaveError) {
+        env.services.notification.add(originalError.message, {
```
[the remark this answers is not in the document]
Comment (aab-odoo) [reply addons/survey/static/src/question_page/question_page_one2many_field.js:41]: > Can't we remove the `notificationService` in the X2ManyField below too?

good catch thanks

**Review by ged-odoo — COMMENTED** [review]
are we sure we want a hook `observeRecord` that does not start with `on` or `use`? maybe would be better to rename it `onRecordChanged` or something?

**Comment by ged-odoo** [comment]
robodoo r+

**Pull request closed without merging**